### PR TITLE
net/freelan: add freelan p2p VPN package to official feed

### PR DESCRIPTION
**Pull request by leggewie**
Signed-off-by: Rolf Leggewie <foss@rolf.leggewie.biz>

Please add freelan p2p VPN software to the official OpenWRT packages feed.  I am working closely with upstream and intend to maintain this package in Debian and OpenWRT in collaboration with @ereOn . Let me know if you have any comments or suggestions for improvement.

### CODE REVIEW

**Comment by champtar** [comment]
Hi @leggewie,
YOU are the PKG_MAINTAINER, not julien.
You should remove the comment at the start (the one about gplv2)
Also if you can switch init script to procd.

I m on my phone, and can't really test until monday
Bye

**Comment by leggewie** [comment]
Thanks, champtar for your comment.

I received most of the work from you and @ereOn who wrote the Makefile with him as maintainer.  Since I only added tweaks and testing I did not want to override him on that decision.  That being said, I'm certainly willing to step up as maintainer if ever the need arose.

**Comment by ereOn** [comment]
@leggewie @champtar When I wrote the Makefile from scratch I indeed put my name as maintainer but clearly it was just to fill the fields, nothing more. I have no issue with having someone else there (actually quite the opposite !).

FreeLAN is an open-source which means it belongs to the community, not just me. Just because I am the original author does not mean I'll ever "override" somebody's decision. Anybody can fork and/or maintain the projet anyway he likes ! (As long as the license terms are respected, obviously).

**Comment by champtar** [comment]
it's not building with latest trunk
```
make[3]: Entering directory `/home/etienne/openwrt-trunk/feeds/packages/net/freelan'
rm -f /home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/stamp/.freelan_installed
(cd /home/etienne/openwrt-trunk/build_dir/target-mips_34kc_musl-1.1.10/freelan-2015-08-01; CC="mips-openwrt-linux-musl-gcc" CXX="mips-openwrt-linux-musl
-g++" CFLAGS="-Os -pipe -mno-branch-likely -mips32r2 -mtune=34kc -fno-caller-saves -fhonour-copts -Wno-error=unused-but-set-variable -msoft-float -Wform
at -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -I/home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl
-1.1.10/usr/lib/libiconv-stub/include -I/home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/lib/libintl-stub/include -Wno-error=cpp -Wno-error=return-type " CXXFLAGS="-Os -pipe -mno-branch-likely -mips32r2 -mtune=34kc -fno-caller-saves -fhonour-copts -Wno-error=unused-but-set-variab
le -msoft-float -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -I/home/etienne/openwrt-trunk/staging_dir
/target-mips_34kc_musl-1.1.10/usr/lib/libiconv-stub/include -I/home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/lib/libintl-stub/
include -Wno-error=cpp -Wno-error=return-type " CPPFLAGS="-I/home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/include -I/home/eti
enne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/include -I/home/etienne/openwrt-trunk/staging_dir/toolchain-mips_34kc_gcc-4.8-linaro_musl-1.
1.10/usr/include -I/home/etienne/openwrt-trunk/staging_dir/toolchain-mips_34kc_gcc-4.8-linaro_musl-1.1.10/include/fortify -I/home/etienne/openwrt-trunk/
staging_dir/toolchain-mips_34kc_gcc-4.8-linaro_musl-1.1.10/include -I/home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/lib/libico
nv-stub/include -I/home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/lib/libintl-stub/include " LDFLAGS="-L/home/etienne/openwrt-t
runk/staging_dir/target-mips_34kc_musl-1.1.10/usr/lib -L/home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/lib -L/home/etienne/openwrt
-trunk/staging_dir/toolchain-mips_34kc_gcc-4.8-linaro_musl-1.1.10/usr/lib -L/home/etienne/openwrt-trunk/staging_dir/toolchain-mips_34kc_gcc-4.8-linaro_m
usl-1.1.10/lib -znow -zrelro -L/home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/lib/libiconv-stub/lib -L/home/etienne/openwrt-tr
unk/staging_dir/target-mips_34kc_musl-1.1.10/usr/lib/libintl-stub/lib " DESTDIR="/home/etienne/openwrt-trunk/build_dir/target-mips_34kc_musl-1.1.10/free
lan-2015-08-01/ipkg-install" CXXFLAGS="$CXXFLAGS -Wno-missing-field-initializer" LIBS="iconv" LINKFLAGS="-L/home/etienne/openwrt-trunk/staging_dir/targe
t-mips_34kc_musl-1.1.10/usr/lib -L/home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/lib -L/home/etienne/openwrt-trunk/staging_dir/too
lchain-mips_34kc_gcc-4.8-linaro_musl-1.1.10/usr/lib -L/home/etienne/openwrt-trunk/staging_dir/toolchain-mips_34kc_gcc-4.8-linaro_musl-1.1.10/lib -znow -
zrelro -L/home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/lib/libiconv-stub/lib -L/home/etienne/openwrt-trunk/staging_dir/target
-mips_34kc_musl-1.1.10/usr/lib/libintl-stub/lib " FREELAN_NO_GIT=1 FREELAN_NO_GIT_VERSION="2015-08-01-3177dd8b21540d8c8d8c3472d85ad00ae4bc39d6" scons pr
efix=/usr "--mode=release" install )
scons: Reading SConscript files ...
scons: done reading SConscript files.
scons: Building targets ...
mips-openwrt-linux-musl-g++ -o build/release/apps/freelan/src/tools.o -c -Os -pipe -mno-branch-likely -mips32r2 -mtune=34kc -fno-caller-saves -fhonour-c
opts -Wno-error=unused-but-set-variable -msoft-float -Wformat -Werror=format-security -fstack-protector -D_FORTIFY_SOURCE=1 -Wl,-z,now -Wl,-z,relro -I/h
ome/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/lib/libiconv-stub/include -I/home/etienne/openwrt-trunk/staging_dir/target-mips_3
4kc_musl-1.1.10/usr/lib/libintl-stub/include -Wno-error=cpp -Wno-error=return-type -Wno-missing-field-initializer --std=c++11 -Wall -Wextra -Werror -ped
antic -Wshadow -O3 -isystem third-party/install/include -I/home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/include -I/home/etien
ne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/include -I/home/etienne/openwrt-trunk/staging_dir/toolchain-mips_34kc_gcc-4.8-linaro_musl-1.1.
10/usr/include -I/home/etienne/openwrt-trunk/staging_dir/toolchain-mips_34kc_gcc-4.8-linaro_musl-1.1.10/include/fortify -I/home/etienne/openwrt-trunk/st
aging_dir/toolchain-mips_34kc_gcc-4.8-linaro_musl-1.1.10/include -I/home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/lib/libiconv
-stub/include -I/home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/lib/libintl-stub/include -DFREELAN_INSTALL_PREFIX="\"/usr\"" -I
build/release/include build/release/apps/freelan/src/tools.cpp
In file included from /home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/include/boost/asio/detail/socket_types.hpp:57:0,
                 from /home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/include/boost/asio/detail/epoll_reactor.hpp:30,
                 from /home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/include/boost/asio/detail/reactor.hpp:21,
                 from /home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/include/boost/asio/detail/impl/task_io_service.ipp:24,
                 from /home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/include/boost/asio/detail/task_io_service.hpp:198,
                 from /home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/include/boost/asio/impl/io_service.hpp:71,
                 from /home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/include/boost/asio/io_service.hpp:767,
                 from /home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/include/boost/asio/basic_io_object.hpp:19,
                 from /home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/include/boost/asio/basic_socket.hpp:20,
                 from /home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/include/boost/asio/basic_datagram_socket.hpp:20,
                 from /home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/usr/include/boost/asio.hpp:21,
                 from build/release/apps/freelan/src/tools.hpp:50,
                 from build/release/apps/freelan/src/tools.cpp:47:
/home/etienne/openwrt-trunk/staging_dir/toolchain-mips_34kc_gcc-4.8-linaro_musl-1.1.10/include/sys/poll.h:1:2: warning: #warning redirecting incorrect #
include <sys/poll.h> to <poll.h> [-Wcpp]
 #warning redirecting incorrect #include <sys/poll.h> to <poll.h>
  ^
cc1plus: error: unrecognized command line option "-Wno-missing-field-initializer" [-Werror]
cc1plus: all warnings being treated as errors
scons: *** [build/release/apps/freelan/src/tools.o] Error 1
scons: building terminated because of errors.
make[3]: *** [/home/etienne/openwrt-trunk/build_dir/target-mips_34kc_musl-1.1.10/freelan-2015-08-01/.configured_yyyyyyyyn] Error 2
make[3]: Leaving directory `/home/etienne/openwrt-trunk/feeds/packages/net/freelan'
make[2]: *** [package/feeds/packages/freelan/compile] Error 2
make[2]: Leaving directory `/home/etienne/openwrt-trunk'
make[1]: *** [/home/etienne/openwrt-trunk/staging_dir/target-mips_34kc_musl-1.1.10/stamp/.package_compile] Error 2
make[1]: Leaving directory `/home/etienne/openwrt-trunk'
make: *** [world] Error 2
```

File: net/freelan/Makefile [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+define Package/freelan/description
+	A free, open-source, multi-platform, highly-configurable and peer-to-peer VPN software, designed to easily connect remote hosts and mainly focused on security and performance.
+endef
+
+TARGET_CFLAGS += -Wno-error=cpp
+TARGET_CFLAGS += -Wno-error=return-type
+
+SCONS_VARS += \
+	CXXFLAGS="$$$$CXXFLAGS -Wno-missing-field-initializer" \
```
Comment (leggewie) [inline net/freelan/Makefile]: unfortunately, this doesn' do what is intended. I've tried numerous other variations, but so far unsuccessfully.

**Comment by leggewie** [comment]
Indeed, it isn't.  I had previously had "successful" builds, but those were spurious when in fact nothing was built (but that successfully, haha).  There seems to be some kind of bug in OpenWRT itself when building a single package, but I'm unable to pinpoint it now.

The problem with freelan is that we need to inject -Wno-missing-field-initializer into CXXFLAGS, but despite a wide variety of attempts, I was so far unsuccessful.  See line 52 in the Makefile.

File: net/freelan/Makefile [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+PKG_SOURCE_VERSION:=3177dd8b21540d8c8d8c3472d85ad00ae4bc39d6
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION)-$(PKG_SOURCE_VERSION).tar.gz
+
+PKG_LICENSE:=GPLv3+
+PKG_LICENSE_FILES:=
+
+PKG_BUILD_PARALLEL:=1
+PKG_USE_MIPS16:=0
+
+PKG_MAINTAINER:="Julien Kauffmann"
```
Comment (leggewie) [inline net/freelan/Makefile]: I will assume role of PKG_MAINTAINER instead of Julien

**Comment by champtar** [comment]
is g++ recent enough (4.8-linaro)?

**Comment by ereOn** [comment]
@champtar It may not be enough. I don't know for this specific `4.8` version but FreeLAN requires a fully C++11 compatible compiler, and I believe `g++` became fully conformant from its `4.9` version.

**Comment by leggewie** [comment]
I do remember the program compiled once in the past.  What do we do now?

**Comment by ereOn** [comment]
@leggewie Could it be that some minor 4.8.x version of g++ was more compatible ? If anything, there should not be that many code fragments that are C++11 specific.

**Comment by sbyx** [comment]
We will eventually switch to 5.x at some point for CC.

**Comment by sbyx** [comment]
@leggewie any new findings on whether this still compiles with latest trunk?

**Comment by champtar** [comment]
trunk recently switched to gcc 5.2
https://dev.openwrt.org/changeset/47492/

**Comment by champtar** [comment]
Hi all,
@leggewie feel free to open a new PR once this build/run (and if you still want freelan in OpenWRT)

**Pull request closed without merging**